diag_manager/fms_diag_buffer.F90

```diff
@@ -0,0 +1,1546 @@
+      if (allocated(this%buffer)) deallocate(this%buffer)
+      if (allocated(this%counter)) deallocate(this%counter)
+    type is (buffer4d_type)
+      if (allocated(this%buffer)) deallocate(this%buffer)
+      if (allocated(this%counter)) deallocate(this%counter)
+    type is (buffer5d_type)
+      if (allocated(this%buffer)) deallocate(this%buffer)
+      if (allocated(this%counter)) deallocate(this%counter)
+  end select
+  if (allocated(this%buffer_id)) deallocate(this%buffer_id)
+  if (allocated(this%count_0d)) deallocate(this%count_0d)
+  if (allocated(this%num_elements)) deallocate(this%num_elements)
+  if (allocated(this%buffer_dims)) deallocate(this%buffer_dims)
+end subroutine flush_buffer
+
+!! -----------Type-specific routines for buffer0-5d
+
+!! allocations could be done in one routine for 0-5d if buffobj is changed to fmsDiagBuffer_class
+!! not sure which approach would be better
+
+!> allocates scalar buffer data to the given buff_type
+subroutine allocate_buffer_0d(this, buff_type, diurnal_samples, field_name)
+  class(buffer0d_type), intent(inout), target :: this !< scalar buffer object
+  class(*),intent(in) :: buff_type !< allocates to the given type, value does not matter
+  integer, optional :: diurnal_samples !< number of diurnal samples, passed in from diag_yaml
+  integer :: n_samples !< number of diurnal samples, defaults to 1
+  character(len=*), optional, intent(in) :: field_name !< optional field name for error output
+  character(len=128)          :: field_name_str
+  if(present(field_name)) then
+    field_name_str = field_name(1:128)
+  else
+    field_name_str = 'N/A'
+  endif
+
+  if(present(diurnal_samples)) then
+    n_samples = diurnal_samples
+  else
+    n_samples = 1
+  endif
+
+  if(allocated(this%buffer)) call mpp_error(FATAL, "allocate_buffer_0d: buffer already allocated for field:"// &
+                                                    field_name_str)
+  select type (buff_type)
+    type is (integer(kind=i4_kind))
+      allocate(integer(kind=i4_kind) :: this%buffer(1))
+      allocate(integer(kind=i4_kind) :: this%counter(1))
+      allocate(integer(kind=i4_kind) :: this%count_0d(n_samples))
+      this%counter = 0
+      this%buffer_type = i4
+    type is (integer(kind=i8_kind))
+      allocate(integer(kind=i8_kind) :: this%buffer(1))
+      allocate(integer(kind=i8_kind) :: this%counter(1))
+      allocate(integer(kind=i8_kind) :: this%count_0d(n_samples))
+      this%counter = 0
+      this%buffer_type = i8
+    type is (real(kind=r4_kind))
+      allocate(real(kind=r4_kind) :: this%buffer(1))
+      allocate(real(kind=r4_kind) :: this%counter(1))
+      allocate(real(kind=r4_kind) :: this%count_0d(n_samples))
+      this%counter = 0
+      this%buffer_type = r4
+    type is (real(kind=r8_kind))
+      allocate(real(kind=r8_kind) :: this%buffer(1))
+      allocate(real(kind=r8_kind) :: this%counter(1))
+      allocate(real(kind=r8_kind) :: this%count_0d(n_samples))
+      this%counter = 0
```

Comment:
You have an implicit casting here which may cause a warning with some compilers
```fortran
      allocate(real(kind=r4_kind) :: this%counter(1))
      this%counter = 0
```
You have to at a minimum add a `.` at the end of the 0.
```fortran
 this%counter = 0.0
```
You can also use a fancier scientific notation with the kind if you want to be specific. 

This applies to the r4 and r8, and all of the other routines that have this.  Sorry I didn't catch it earlier.

Reply:
fixed, I added the fancy notation. I moved count_0d as well so it's allocated with the buffer.